### PR TITLE
New Script: Apache Guacamole

### LINKED ISSUE #656: Apache Guacamole
https://github.com/community-scripts/ProxmoxVE/discussions/29

Application Name
Apache Guacamole

Website
https://guacamole.apache.org/

Description
Apache Guacamole is a clientless remote desktop gateway. It supports standard protocols like VNC, RDP, and SSH.
We call it clientless because no plugins or client software are required.
Thanks to HTML5, once Guacamole is installed on a server, all you need to access your desktops is a web browser.

### PR DESCRIPTION
**Pull request by michelroegl-brunner**
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description
New Script for Apache Guacamole. 
At the moment i have no Update funktion implemented, it seems it can lead to many issues to do this automaticly, but i will figure it out and create a new PR for this feature.

Fixes #656 

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [X] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [X] Testing performed (I have tested my changes, ensuring everything works as expected)
- [X] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)
Provide any additional context or screenshots about the feature or fix here.

## Related Pull Requests / Discussions

If there are other pull requests or discussions related to this change, please link them here:
- Related PR #

### CODE REVIEW

File: json/apache-guacamole.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "notes": [
+    {
+      "text": "After installation you can reach the Management Dashboard with $IP:8080/guacamole"
+      "type": "info"
+    }
+    {
+      "text": "Default Username is guacadmin, Password is guacadmin"
+      "type": "info"
+    }
```
Comment (BramSuurdje) [inline json/apache-guacamole.json:41]: you can remove this after changing the default_credentials

File: json/apache-guacamole.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        "hdd": 4,
+        "os": "Debian",
+        "version": "12"
+      }
+    }
+  ],
+  "default_credentials": {
+    "username": null,
+    "password": null
+  },
```
Comment (BramSuurdje) [inline json/apache-guacamole.json:32]: "username": "guacadmin",
"password": "guacadmin"

File: json/apache-guacamole.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  "name": "Apache Guacamole",
+  "slug": "apache-guacamole",
+  "categories": [
+    11
+  ],
+  "date_created": "2024-12-02",
+  "type": "ct",
+  "updateable": false,
+  "privileged": false,
+  "interface_port": null,
```
Comment (BramSuurdje) [inline json/apache-guacamole.json:11]: "interface_port": 8080,

**Review by bramsuurdje — CHANGES_REQUESTED** [review]

File: ct/apache-guacamole.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
+source <(curl -s https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/dev/misc/build.func)
```
Comment (MickLesk) [inline ct/apache-guacamole.sh:2]: Nope ^^

File: install/apache-guacamole-install.sh [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+   libpango1.0-dev \
+   libswscale-dev \
+   libavcodec-dev \
+   libavutil-dev \
+   libavformat-dev \
+   mariadb-server \
+   default-jdk
+msg_ok "Installed Dependencies"
+
+msg_info "Install Tomcat"
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:44]: msg_info "Setup Apache Tomcat"

File: install/apache-guacamole-install.sh [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+mkdir /opt/${APPLICATION}
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:45]: Please use fixed values, not Variables

File: install/apache-guacamole-install.sh [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+mkdir /opt/${APPLICATION}/tomcat9
+$STD wget https://dlcdn.apache.org/tomcat/tomcat-9/v9.0.97/bin/apache-tomcat-9.0.97.tar.gz
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:47]: Std not needed, wget -q

File: install/apache-guacamole-install.sh [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+$STD tar xzf apache-tomcat-9.0.97.tar.gz -C /opt/${APPLICATION}/tomcat9 --strip-components=1
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:48]: Use static values

File: install/apache-guacamole-install.sh [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+useradd -r -d /opt/tomcat9 -s /bin/false tomcat
+chown -R tomcat: /opt/${APPLICATION}/tomcat9/{logs,temp,webapps,work}
+chown -R :tomcat /opt/${APPLICATION}/tomcat9/
+chmod -R g+r /opt/${APPLICATION}/tomcat9/conf
+chmod g+x /opt/${APPLICATION}/tomcat9/conf
+JAVA_HOME=$(update-alternatives --query javadoc | grep Value: | head -n1 | sed 's/Value: //' | sed 's@bin/javadoc$@@')
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:54]: Is java_home already Set?

File: install/apache-guacamole-install.sh [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+   default-jdk
+msg_ok "Installed Dependencies"
+
+msg_info "Install Tomcat"
+mkdir /opt/${APPLICATION}
+mkdir /opt/${APPLICATION}/tomcat9
+$STD wget https://dlcdn.apache.org/tomcat/tomcat-9/v9.0.97/bin/apache-tomcat-9.0.97.tar.gz
+$STD tar xzf apache-tomcat-9.0.97.tar.gz -C /opt/${APPLICATION}/tomcat9 --strip-components=1
+useradd -r -d /opt/tomcat9 -s /bin/false tomcat
+chown -R tomcat: /opt/${APPLICATION}/tomcat9/{logs,temp,webapps,work}
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:50]: User needed? Why not run as root Default?

File: install/apache-guacamole-install.sh [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+Group=tomcat
+UMask=0007
+RestartSec=10
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
+EOF
+$STD systemctl daemon-reload
+$STD systemctl enable --now tomcat
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:83]: Without Std, -q 
Should be below after all Installations in a Special Section 

msg_info "Setup Services"
.
.
.

File: install/apache-guacamole-install.sh [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+User=tomcat
+Group=tomcat
+UMask=0007
+RestartSec=10
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
+EOF
+$STD systemctl daemon-reload
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:82]: Not needed

File: install/apache-guacamole-install.sh [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+$STD systemctl enable --now tomcat
+msg_ok "Installed Tomcat"
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:84]: Setup Apache Tomcat

File: install/apache-guacamole-install.sh [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+msg_info "Installing Guacamole"
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:86]: Setup Apache Guacamole

File: install/apache-guacamole-install.sh [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+mkdir -p /etc/guacamole/{extensions,lib}
+REALESE_SERVER=$(curl -sL https://api.github.com/repos/apache/guacamole-server/tags | jq -r '.[0].name')
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:88]: Spell issue Overall

File: install/apache-guacamole-install.sh [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+mkdir /opt/${APPLICATION}/apache-guacamole-server-${REALESE_SERVER}
+$STD wget -O apache-guacamole-server-${REALESE_SERVER} https://api.github.com/repos/apache/guacamole-server/tarball/refs/tags/${REALESE_SERVER}
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:90]: STD not needed, -q

File: install/apache-guacamole-install.sh [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+$STD tar -xvf apache-guacamole-server-${REALESE_SERVER} -C /opt/${APPLICATION}/apache-guacamole-server-${REALESE_SERVER} --strip-components=1
+cd /opt/${APPLICATION}/apache-guacamole-server-${REALESE_SERVER}
+$STD autoreconf -fi
+$STD ./configure --with-init-dir=/etc/init.d --enable-allow-freerdp-snapshots
+$STD make
+$STD make install
+$STD ldconfig
+$STD systemctl daemon-reload
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:98]: Why?

File: install/apache-guacamole-install.sh [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+$STD systemctl enable --now guacd
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:99]: Without Std, -q 
Should be below after all Installations in a Special Section 

msg_info "Setup Services"
.
.
.

File: install/apache-guacamole-install.sh [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+REALESE_CLIENT=$(curl -sL https://api.github.com/repos/apache/guacamole-client/tags | jq -r '.[0].name')
+$STD wget -O /opt/${APPLICATION}/tomcat9/webapps/guacamole.war https://downloads.apache.org/guacamole/${REALESE_CLIENT}/binary/guacamole-${REALESE_CLIENT}.war
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:101]: Without Std, -q

File: install/apache-guacamole-install.sh [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+$STD systemctl restart guacd tomcat
+msg_ok "Installed Guacamole"
+
+msg_info "Creating Service"
+cd /root
+$STD wget https://dev.mysql.com/get/Downloads/Connector-J/mysql-connector-java-8.0.26.tar.gz
+$STD tar -xf mysql-connector-java-8.0.26.tar.gz
+mv mysql-connector-java-8.0.26/mysql-connector-java-8.0.26.jar /etc/guacamole/lib/
+$STD wget https://downloads.apache.org/guacamole/1.5.5/binary/guacamole-auth-jdbc-1.5.5.tar.gz
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:110]: -q not Std

File: install/apache-guacamole-install.sh [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+$STD tar -xf guacamole-auth-jdbc-1.5.5.tar.gz
+mv guacamole-auth-jdbc-1.5.5/mysql/guacamole-auth-jdbc-mysql-1.5.5.jar /etc/guacamole/extensions/
+DB_NAME=guacamole_db
```
Comment (MickLesk) [inline install/apache-guacamole-install.sh:113]: Here is an Single Section

msg_info "Setup Database....... Or something else.

**Review by MickLesk — COMMENTED** [review]
Phew, the script is terrible :smile: spelling mistakes and various violations of the conventions. Have a look at other current scripts. I just got a real headache trying to follow the script. 

The following:
1. installing dependencies 
2. setup database
3. setup Tomcat
4. setup Guacamole
5. creating service
6. cleanup

Use STD only when absolutely necessary, wget, curl, unzip and so on support -q 

Then please do not use variable paths, but fixed paths.

**Comment by michelroegl-brunner** [comment]
> Phew, the script is terrible 😄 spelling mistakes and various violations of the conventions. Have a look at other current scripts. I just got a real headache trying to follow the script.
> 
> The following:
> 
> 1. installing dependencies
> 2. setup database
> 3. setup Tomcat
> 4. setup Guacamole
> 5. creating service
> 6. cleanup
> 
> Use STD only when absolutely necessary, wget, curl, unzip and so on support -q
> 
> Then please do not use variable paths, but fixed paths.

First of all thank you for your honesty :D I´m still trying to get the gist of this codebase an highly appreciate the time and effort you put in this reviews and the whole project. 
This was a big script for me and i work in all the changes tomorrow!

File: install/apache-guacamole-install.sh (hunk 9 above)
Comment (michelroegl-brunner) [reply install/apache-guacamole-install.sh:54]: This Variable is used to set the correct Java_home in the tomcat service.

File: ct/apache-guacamole.sh (hunk 4 above)
Comment (michelroegl-brunner) [reply ct/apache-guacamole.sh:2]: Good catch^^

File: install/apache-guacamole-install.sh (hunk 10 above)
Comment (michelroegl-brunner) [reply install/apache-guacamole-install.sh:50]: Yep i just testet it. Tomcat will not start under root, so the user is necessary.

**Comment by michelroegl-brunner** [comment]
Updated to account for the changes. Test it on my dev branch with `bash -c "$(wget -qLO -https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/DEV/ct/apache-guacamole.sh)"`

**Comment by MickLesk** [comment]
if you have no update function, check adguard.sh (ct), otherwise the user will not get any feedback

**Comment by MickLesk** [comment]
Error: ENOENT: no such file or directory, stat '/ct/apache-guacamole.sh'

json issue, idk why

**Comment by michelroegl-brunner** [comment]
> Error: ENOENT: no such file or directory, stat '/ct/apache-guacamole.sh'
> 
> json issue, idk why

I´ve also seen this. Maybe an issue with the frontend workflow? It probably checks the main branch and there can never be such a file when its a PR with a new script.

I have also added the update_function() as requested.

**Comment by MickLesk** [comment]
@havardthom can you check the workflow? i dont see any json issue here? :smile:

**Comment by MickLesk** [comment]
ive see the issue now.
its an unneeded / in script 

@BramSuurdje can you check it?
image

json-editor (https://community-scripts.github.io/ProxmoxVE/json-editor):
image

**Comment by MickLesk** [comment]
The script has now been finalized. I have optimized it a little and tested it in the DEV branch. The script is a bit special because we have to use various sources here. 

can be tested here:
```bash
bash -c "$(wget -qLO - https://raw.githubusercontent.com/community-scripts/ProxmoxVE/refs/heads/develop/ct/apache-guacamole.sh)"
```

@CrazyWolf13 | @IVRYSimon | @trucke 
Can you check it? I don't really know the use-case, I just see an empty board where I have to add connections. idk what I have to do :D

**Comment by joe-heaton** [comment]
Happy to test if that's ok? 
Have used this app before so something to base off 👍🏻

**Comment by michelroegl-brunner** [comment]
> Happy to test if that's ok? 
> Have used this app before so something to base off 👍🏻

Yes please, highly appreciated :)

**Comment by joe-heaton** [comment]
> > Happy to test if that's ok? 
> > Have used this app before so something to base off 👍🏻
> 
> Yes please, highly appreciated :)

Installed great, little long but my internet is below par so probably that.

Tested working for both VNC and RDP connections to a Windows Machine

**Review by MickLesk — APPROVED** [review]

**Review by CrazyWolf13 — COMMENTED** [review]
@michelroegl-brunner @MickLesk  Any reason this contains Snipe-IT ?

File: install/apache-guacamole-install.sh [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+DB_PASS=$(openssl rand -base64 18 | tr -dc 'a-zA-Z0-9' | head -c13)
+mysql -u root -e "CREATE DATABASE $DB_NAME;"
+mysql -u root -e "CREATE USER '$DB_USER'@'localhost' IDENTIFIED WITH mysql_native_password AS PASSWORD('$DB_PASS');"
+mysql -u root -e "GRANT ALL ON $DB_NAME.* TO '$DB_USER'@'localhost'; FLUSH PRIVILEGES;"
+{
+    echo "SnipeIT-Credentials"
+    echo "SnipeIT Database User: $DB_USER"
+    echo "SnipeIT Database Password: $DB_PASS"
+    echo "SnipeIT Database Name: $DB_NAME"
+} >> ~/guacamole.creds
```
Comment (CrazyWolf13) [inline install/apache-guacamole-install.sh:87]: @michelroegl-brunner @MickLesk Any reason this contains Snipe-IT ?

Comment (MickLesk) [reply install/apache-guacamole-install.sh:87]: I think Copy Paste issue. I Update it

**Comment by trucke** [comment]
> The script has now been finalized. I have optimized it a little and tested it in the DEV branch. The script is a bit special because we have to use various sources here.
> 
> can be tested here:
> 
> ```shell
> bash -c "$(wget -qLO - https://raw.githubusercontent.com/community-scripts/ProxmoxVE/refs/heads/develop/ct/apache-guacamole.sh)"
> ```
> 
> @CrazyWolf13 | @IVRYSimon | @trucke Can you check it? I don't really know the use-case, I just see an empty board where I have to add connections. idk what I have to do :D
> 
> Edit: Please only test, if its okay, you can delete the lxc after this, and we merge to main

sorry for my late reply, unfortunately i'm sick in bed at the moment. i will test it as soon as i can.
thanks guys for your hard and diligent work.

**Comment by CrazyWolf13** [comment]
@MickLesk Sorry for the delay in testing, had to get my pve cluster fixed first.

TLDR: For me  it's compledtely not working!

For me adding a remote connection, for example VNC, immediately fails:
grafik
A connection with the exact same settings from my prod guacamole works just fine.
SSH connections fail as well immediately, so I guess there must be some sort of problem between guacd and guac frontend.

Normally this command enables guacd with debug logs to cli, however absolutely 0 logs appear here:
```
systemctl stop guacd
/usr/local/sbin/guacd -L debug -f
```

Maybe we can use inspiration at this code: https://github.com/itiligent/Easy-Guacamole-Installer

**Comment by CrazyWolf13** [comment]
@MickLesk  @joe-heaton  @michelroegl-brunner @trucke 

Was any of you able to get guacamole working?
